pkg/kubectl/role.go

```diff
@@ -0,0 +1,304 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubectl
+
+import (
+	"fmt"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/kubernetes/pkg/apis/rbac"
+)
+
+// RoleGeneratorV1 supports stable generation of a role.
+type RoleGeneratorV1 struct {
+	// Name of role (required)
+	Name string
+	// Verbs is a list of Verbs that apply to ALL the ResourceKinds contained in this rule (required)
+	Verbs []string
+	// Resources is a list of resources this rule applies to (required)
+	Resources []string
+	// APIGroups is the name of the APIGroup that contains the resources (optional)
+	APIGroups []string
+	// ResourceNames is an optional white list of names that the rule applies to (optional)
+	ResourceNames []string
+	// Mapper is used to validate resources.
+	Mapper meta.RESTMapper
+}
+
+// Ensure it supports the generator pattern that uses parameter injection.
+var _ Generator = &RoleGeneratorV1{}
+
+// Ensure it supports the generator pattern that uses parameters specified during construction.
+var _ StructuredGenerator = &RoleGeneratorV1{}
+
+// Valid verb list for validation.
+var validVerbs []string = []string{"*", "get", "post", "put", "delete", "list", "create", "update", "patch", "watch", "proxy", "redirect", "deletecollection"}
```

Comment:
split into valid resource version and valid non-resource verbs.  Resource verbs should probably be *, get, list, watch, update, patch, delete, deletecollection